homeassistant/components/smartthings/light.py

```diff
@@ -0,0 +1,204 @@
+
+
+def is_light(device):
+    """Determine if the device should be represented as a light."""
+    from pysmartthings import Capability
+
+    # Must be able to be turned on/off.
+    if Capability.switch not in device.capabilities:
+        return False
+    # Not a fan (which might also have switch_level)
+    if Capability.fan_speed in device.capabilities:
+        return False
+    # Must have one of these
+    light_capabilities = [
+        Capability.color_control,
+        Capability.color_temperature,
+        Capability.switch_level
+    ]
+    if any(capability in device.capabilities
+           for capability in light_capabilities):
+        return True
+    return False
+
+
+def convert_scale(value, value_scale, target_scale, round_digits=4):
+    """Convert a value to a different scale."""
+    return round(value * target_scale / value_scale, round_digits)
+
+
+class SmartThingsLight(SmartThingsEntity, Light):
+    """Define a SmartThings Light."""
+
+    def __init__(self, device):
+        """Initialize an SmartThingsLight."""
+        SmartThingsEntity.__init__(self, device)
+        self._brightness = 0
+        self._hs_color = (0.0, 0.0)
+        self._supported_features = 0
+        self._color_temp = 0
+
+    async def async_turn_on(self, **kwargs) -> None:
+        """Turn the light on."""
+        tasks = []
+        # Color temperature
+        if self._supported_features & SUPPORT_COLOR_TEMP \
+           and ATTR_COLOR_TEMP in kwargs:
+            tasks.append(self.set_color_temp(
+                kwargs[ATTR_COLOR_TEMP]))
+        # Color
+        if self._supported_features & SUPPORT_COLOR \
+           and ATTR_HS_COLOR in kwargs:
+            tasks.append(self.set_color(
+                kwargs[ATTR_HS_COLOR]))
+        if tasks:
+            # Set temp/color first
+            await asyncio.gather(*tasks)
+
+        # Switch/brightness/transition
+        if self._supported_features & SUPPORT_BRIGHTNESS \
+           and ATTR_BRIGHTNESS in kwargs:
+            await self.set_level(
+                kwargs[ATTR_BRIGHTNESS],
+                kwargs.get(ATTR_TRANSITION, 0))
+        else:
+            await self._device.switch_on(set_status=True)
+
+        # State is set optimistically in the commands above, therefore update
+        # the entity state ahead of receiving the confirming push updates
+        self.async_schedule_update_ha_state(True)
+
+    async def async_turn_off(self, **kwargs) -> None:
+        """Turn the light off."""
+        # Switch/transition
+        if self._supported_features & SUPPORT_TRANSITION \
+                and ATTR_TRANSITION in kwargs:
+            await self.set_level(0, int(kwargs[ATTR_TRANSITION]))
+        else:
+            await self._device.switch_off(set_status=True)
+
+        # State is set optimistically in the commands above, therefore update
+        # the entity state ahead of receiving the confirming push updates
+        self.async_schedule_update_ha_state(True)
```

Comment:
Will the device know about the new state when we access state in `async_update`?

Reply:
Yes -- per the comment above the line, it's set optimistically.